README.md
# Flutter location_provider Plugin

[![pub package](https://img.shields.io/pub/v/geocoding.svg)](https://pub.dartlang.org/packages/geocoding)
[![License: MIT](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)

A Flutter location_provider plugin which provides easy address and latlong and reverse-geocoding features.

**Important**:

1. This plugin uses the free Geocoding services provided by the iOS and Android platforms. This means that there are restrictions to their use. More information can be found in the [Apple documentation for iOS](https://developer.apple.com/documentation/corelocation/clgeocoder) and the [Google documentation for Android](https://developer.android.com/reference/android/location/Geocoder).
   When a `PlatformException(IO_ERROR, ...)` gets thrown, most of the times it means that the rate limit has been reached.
2. The availability of the Google Play Services depends on your country. If your country doesn't support a connection with the Google Play Services, you'll need to try a VPN to establish a connection. For more information about how to work with Google Play Services visit the following link: https://developers.google.com/android/guides/overview

## Usage

To use this plugin, please follow the installation guide on the [official geocoding plugin page](https://pub.dev/packages/geocoding/install).

> **NOTE:** This plugin relies on the AndroidX version of the Android Support Libraries. This means you need to make sure your Android project is also upgraded to support AndroidX. Detailed instructions can be found [here](https://flutter.dev/docs/development/packages-and-plugins/androidx-compatibility).
>
>The TL;DR version is:
>
>1. Add the following to your "gradle.properties" file:
>
>```
>android.useAndroidX=true
>android.enableJetifier=true
>```
>2. Make sure you set the `compileSdkVersion` in your "android/app/build.gradle" file to 31:
>
>```
>android {
>  compileSdkVersion 31
>
>  ...
>}
>```
>3. Make sure you replace all the `android.` dependencies to their AndroidX counterparts (a full list can be found [Android migration guide](https://developer.android.com/jetpack/androidx/migrate)).

## API

To translate an address into latitude and longitude coordinates you can use the `placemarkFromAddress` method:

``` dart
import 'package:location_provider/location_provider.dart';
final lat = await LocationHelper.getLatitudeFromAddress(address);
final lng = await LocationHelper.getLongitudeFromAddress(address);
```

If you want to translate latitude and longitude coordinates into an address you can use the `placemarkFromCoordinates` method:

``` dart
import 'package:location_provider/location_provider.dart';

final address = await LocationHelper.getLocationFullAddress(lat, lan);
```
## Example (Recording)

https://user-images.githubusercontent.com/33664603/215685551-7dc56191-74d0-453d-b93c-05cc9dcdf8fb.mov

## Issues

Please file any issues, bugs or feature requests as an issue, you can contact us at <rahulkushwaha482@gmail.com>.

## Author

This geocoding plugin for Flutter is developed by [RahulKushwaha](https://in.linkedin.com/in/rahulkushwaha482).


CHANGELOG.md
## 0.0.1

* Intial release of location_provider.

## 0.0.2

* Code formatted.

## 0.0.3

* Example added.

example/home.dart
import 'package:flutter/material.dart';
import 'package:location_provider/location_provider.dart';

class HomePage extends HookWidget {
  void _updateAddress({
    required ValueNotifier<TextEditingValue> latitudeController,
    required ValueNotifier<TextEditingValue> longitudeController,
    required ValueNotifier<String> address,
  }) async {
    final lat = double.tryParse(latitudeController.value.text);
    final lan = double.tryParse(longitudeController.value.text);
    if (lat != null && lan != null) {
      final addre = await LocationHelper.getLocationFullAddress(lat, lan);
      address.value = addre.toString();
    } else {
      address.value = 'address';
    }
  }

  void _updateLatLng({
    required ValueNotifier<TextEditingValue> addressController,
    required ValueNotifier<String> lat,
    required ValueNotifier<String> lang,
  }) async {
    final address = (addressController.value.text);

    if (address != null) {
      final addre = await LocationHelper.getLatitudeFromAddress(address);
      final lng = await LocationHelper.getLongitudeFromAddress(address);
      lat.value = addre.toString();
      lang.value = lng.toString();
    } else {
      lat.value = '';
      lang.value = '';
    }
  }

  @override
  Widget build(BuildContext context) {
    final address = useState('Address');
    final lat = useState('');
    final lan = useState('');

    final latitudeController = useTextEditingController();
    final longitudeController = useTextEditingController();
    final addressController = useTextEditingController();
    return SafeArea(
      child: Scaffold(
        body: SingleChildScrollView(
          child: Column(
            children: [
              const SizedBox(
                height: 50,
              ),
              const Text('Get Address form latlong'),
              TextField(
                  controller: latitudeController,
                  decoration: const InputDecoration(labelText: 'Latitude'),
                  keyboardType: TextInputType.number,
                  onChanged: (_) => _updateAddress(
                      latitudeController: latitudeController,
                      longitudeController: longitudeController,
                      address: address)),
              TextField(
                  controller: longitudeController,
                  decoration: const InputDecoration(labelText: 'Longitude'),
                  keyboardType: TextInputType.number,
                  onChanged: (_) => _updateAddress(
                      latitudeController: latitudeController,
                      longitudeController: longitudeController,
                      address: address)),
              Text(
                address.value.toString(),
                style: const TextStyle(
                  fontSize: 15,
                ),
              ),
              const Text('Get LatLong form Address'),
              TextField(
                  controller: addressController,
                  decoration: const InputDecoration(labelText: 'Address'),
                  keyboardType: TextInputType.text,
                  onChanged: (_) => _updateLatLng(
                      addressController: addressController,
                      lat: lat,
                      lang: lan)),
              Text('${lat.value.toString()},${lan.value.toString()}'),
            ],
          ),
        ),
      ),
    );
  }
}


lib/src/location_helper.dart
import 'package:geocoding/geocoding.dart';

class LocationHelper {
  static Future<String> getLocationFullAddress(double lat, double lng) async {
    List<Placemark> placemarks = await placemarkFromCoordinates(lat, lng);
    String platformVersion =
        '${placemarks[0].subLocality},${placemarks[0].locality},${placemarks[0].country}';
    return platformVersion;
  }

  static Future<String> getSubLocality(double lat, double lng) async {
    List<Placemark> placemarks = await placemarkFromCoordinates(lat, lng);
    final subLocality = '${placemarks[0].subLocality}';

    return subLocality;
  }

  static Future<String> getLocality(double lat, double lng) async {
    List<Placemark> placemarks = await placemarkFromCoordinates(lat, lng);
    final locality = '${placemarks[0].locality}';
    return locality;
  }

  static Future<String> getCountry(double lat, double lng) async {
    List<Placemark> placemarks = await placemarkFromCoordinates(lat, lng);
    final locality = '${placemarks[0].country}';
    return locality;
  }

  static Future<String> getLatitudeFromAddress(String address) async {
    List<Location> locations = await locationFromAddress(address);
    String latLong = '${locations.first.latitude}';
    return latLong;
  }

  static Future<String> getLongitudeFromAddress(String address) async {
    List<Location> locations = await locationFromAddress(address);
    String latLong = '${locations.first.longitude}';
    return latLong;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_hooks/flutter_hooks.dart';
import 'package:location_provider/location_provider.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home:  HomePage(),
    );
  }
}

class HomePage extends HookWidget {
  void _updateAddress({
    required ValueNotifier<TextEditingValue> latitudeController,
    required ValueNotifier<TextEditingValue> longitudeController,
    required ValueNotifier<String> address,
  }) async {
    final lat = double.tryParse(latitudeController.value.text);
    final lan = double.tryParse(longitudeController.value.text);
    if (lat != null && lan != null) {
      final addre = await LocationHelper.getLocationFullAddress(lat, lan);
      address.value = addre.toString();
    } else {
      address.value = 'address';
    }
  }

  void _updateLatLng({
    required ValueNotifier<TextEditingValue> addressController,
    required ValueNotifier<String> lat,
    required ValueNotifier<String> lang,
  }) async {
    final address = (addressController.value.text);

    if (address != null) {
      final addre = await LocationHelper.getLatitudeFromAddress(address);
      final lng = await LocationHelper.getLongitudeFromAddress(address);
      lat.value = addre.toString();
      lang.value = lng.toString();
    } else {
      lat.value = '';
      lang.value = '';
    }
  }

  @override
  Widget build(BuildContext context) {
    final address = useState('Address');
    final lat = useState('');
    final lan = useState('');

    final latitudeController = useTextEditingController();
    final longitudeController = useTextEditingController();
    final addressController = useTextEditingController();
    return SafeArea(
      child: Scaffold(
        body: SingleChildScrollView(
          child: Column(
            children: [
              const SizedBox(
                height: 50,
              ),
              const Text('Get Address form latlong'),
              TextField(
                  controller: latitudeController,
                  decoration: const InputDecoration(labelText: 'Latitude'),
                  keyboardType: TextInputType.number,
                  onChanged: (_) => _updateAddress(
                      latitudeController: latitudeController,
                      longitudeController: longitudeController,
                      address: address)),
              TextField(
                  controller: longitudeController,
                  decoration: const InputDecoration(labelText: 'Longitude'),
                  keyboardType: TextInputType.number,
                  onChanged: (_) => _updateAddress(
                      latitudeController: latitudeController,
                      longitudeController: longitudeController,
                      address: address)),
              Text(
                address.value.toString(),
                style: const TextStyle(
                  fontSize: 15,
                ),
              ),
              const Text('Get LatLong form Address'),
              TextField(
                  controller: addressController,
                  decoration: const InputDecoration(labelText: 'Address'),
                  keyboardType: TextInputType.text,
                  onChanged: (_) => _updateLatLng(
                      addressController: addressController,
                      lat: lat,
                      lang: lan)),
              Text('${lat.value.toString()},${lan.value.toString()}'),
            ],
          ),
        ),
      ),
    );
  }
}



